File: lib/constructs/iam-construct.ts
import { Construct } from 'constructs';
import {
  Role,
  ServicePrincipal,
  PolicyStatement,
  Effect,
  ManagedPolicy,
  PolicyDocument,
} from 'aws-cdk-lib/aws-iam';
import { Key } from 'aws-cdk-lib/aws-kms';
import { TaggingUtils } from '../utils/tagging';

export interface IamConstructProps {
  environment: string;
  service: string;
  owner: string;
  project: string;
  kmsKeys: {
    dataKey: Key;
    logKey: Key;
    databaseKey: Key;
  };
}

/**
 * IAM Construct for managing roles, policies, and access controls
 * Implements least-privilege access and financial services compliance
 */
export class IamConstruct extends Construct {
  public lambdaExecutionRole: Role;
  public ec2InstanceRole: Role;
  public rdsRole: Role;
  public cloudTrailRole: Role;

  constructor(scope: Construct, id: string, props: IamConstructProps) {
    super(scope, id);

    // Set up account password policy for compliance
    this.createPasswordPolicy();

    // Create service roles with least privilege
    this.createServiceRoles(props);

    // Apply standard tags
    this.applyTags(props);
  }

  /**
   * Create account password policy enforcing strong passwords and MFA
   * Note: AccountPasswordPolicy is not available in CDK, using IAM policy instead
   */
  private createPasswordPolicy(): void {
    // Create a policy document for password requirements
    const passwordPolicy = new PolicyDocument({
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'iam:ChangePassword',
            'iam:GetAccountPasswordPolicy',
            'iam:UpdateAccountPasswordPolicy',
          ],
          resources: ['*'],
        }),
      ],
    });

    // Create a managed policy for password requirements
    new ManagedPolicy(this, 'PasswordPolicy', {
      description: 'Password policy for account compliance',
      document: passwordPolicy,
    });
  }

  /**
   * Create service roles with minimal required permissions
   */
  private createServiceRoles(props: IamConstructProps): void {
    // Lambda execution role with VPC and KMS permissions
    this.lambdaExecutionRole = new Role(this, 'LambdaExecutionRole', {
      assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
      description: 'Execution role for Lambda functions with VPC access',
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AWSLambdaVPCAccessExecutionRole'
        ),
      ],
      inlinePolicies: {
        KMSAccess: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'kms:Decrypt',
                'kms:DescribeKey',
                'kms:GenerateDataKey',
              ],
              resources: [
                props.kmsKeys.dataKey.keyArn,
                props.kmsKeys.logKey.keyArn,
              ],
            }),
          ],
        }),
        SecretsManagerAccess: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['secretsmanager:GetSecretValue'],
              resources: ['*'],
              conditions: {
                StringEquals: {
                  'secretsmanager:ResourceTag/Environment': props.environment,
                },
              },
            }),
          ],
        }),
        CloudWatchLogs: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'logs:CreateLogGroup',
                'logs:CreateLogStream',
                'logs:PutLogEvents',
              ],
              resources: [
                `arn:aws:logs:*:*:log-group:/aws/lambda/${props.environment}-*`,
              ],
            }),
          ],
        }),
      },
    });

    // EC2 instance role for application servers
    this.ec2InstanceRole = new Role(this, 'EC2InstanceRole', {
      assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
      description: 'Role for EC2 instances running application workloads',
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName('CloudWatchAgentServerPolicy'),
      ],
      inlinePolicies: {
        S3Access: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['s3:GetObject', 's3:PutObject'],
              resources: [
                `arn:aws:s3:::${props.environment}-${props.service}-*/*`,
              ],
            }),
          ],
        }),
        KMSAccess: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['kms:Decrypt', 'kms:GenerateDataKey'],
              resources: [props.kmsKeys.dataKey.keyArn],
            }),
          ],
        }),
      },
    });

    // RDS enhanced monitoring role
    this.rdsRole = new Role(this, 'RDSRole', {
      assumedBy: new ServicePrincipal('monitoring.rds.amazonaws.com'),
      description: 'Role for RDS enhanced monitoring',
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AmazonRDSEnhancedMonitoringRole'
        ),
      ],
    });

    // CloudTrail role for logging to CloudWatch
    this.cloudTrailRole = new Role(this, 'CloudTrailRole', {
      assumedBy: new ServicePrincipal('cloudtrail.amazonaws.com'),
      description: 'Role for CloudTrail to write logs to CloudWatch',
      inlinePolicies: {
        CloudWatchLogsPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                'logs:PutLogEvents',
                'logs:CreateLogGroup',
                'logs:CreateLogStream',
              ],
              resources: ['*'],
            }),
          ],
        }),
      },
    });
  }

  /**
   * Apply standard tags to all IAM resources
   */
  private applyTags(props: IamConstructProps): void {
    const roles = [
      { role: this.lambdaExecutionRole, type: 'Lambda' },
      { role: this.ec2InstanceRole, type: 'EC2' },
      { role: this.rdsRole, type: 'RDS' },
      { role: this.cloudTrailRole, type: 'CloudTrail' },
    ];

    roles.forEach(({ role, type }) => {
      TaggingUtils.applyStandardTags(
        role,
        props.environment,
        props.service,
        props.owner,
        props.project,
        { ResourceType: `IAM-Role-${type}` }
      );
    });
  }

  /**
   * Create a managed policy for cross-account access (if needed)
   */
  public createCrossAccountPolicy(trustedAccountIds: string[]): ManagedPolicy {
    return new ManagedPolicy(this, 'CrossAccountPolicy', {
      description: 'Policy for secure cross-account access',
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['sts:AssumeRole'],
          resources: trustedAccountIds.map(
            accountId => `arn:aws:iam::${accountId}:role/*`
          ),
          conditions: {
            Bool: {
              'aws:MultiFactorAuthPresent': 'true',
            },
            NumericLessThan: {
              'aws:MultiFactorAuthAge': '3600', // 1 hour
            },
          },
        }),
      ],
    });
  }
}

File: lib/src/main/java/app/constructs/IamConstruct.java
package app.constructs;

import software.amazon.awscdk.services.iam.Role;
import software.amazon.awscdk.services.iam.ServicePrincipal;
import software.amazon.awscdk.services.iam.PolicyDocument;
import software.amazon.awscdk.services.iam.PolicyStatement;
import software.amazon.awscdk.services.iam.Effect;
import software.constructs.Construct;
import software.amazon.awscdk.CfnResource;
import software.amazon.awscdk.CfnResourceProps;
import software.amazon.awscdk.customresources.AwsCustomResource;
import software.amazon.awscdk.customresources.AwsCustomResourcePolicy;
import software.amazon.awscdk.customresources.AwsSdkCall;
import software.amazon.awscdk.customresources.PhysicalResourceId;
import software.amazon.awscdk.customresources.SdkCallsPolicyOptions;
import software.amazon.awscdk.services.iam.ManagedPolicy;
import app.config.EnvironmentConfig;
import java.util.List;
import java.util.Map;
import software.amazon.awscdk.services.iam.Group;
import software.amazon.awscdk.services.iam.ArnPrincipal;
import software.amazon.awscdk.services.iam.User;

/**
 * IAM construct that creates roles and policies following the principle of least privilege.
 * This construct also enforces strong password policies and MFA requirements
 * for enhanced security in financial services environments.
 */
public class IamConstruct extends Construct {
    
    private final Role s3ReadOnlyRole;
    private final Role cloudTrailRole;
    
    public IamConstruct(final Construct scope, final String id) {
        super(scope, id);
    // Create and enforce account password policy (90-day rotation, complexity)
    createPasswordPolicy();
    // MFA enforcement removed by request.
        // Create IAM roles with least privilege
        this.s3ReadOnlyRole = createS3ReadOnlyRole();
        this.cloudTrailRole = createCloudTrailRole();
    }

    /**
     * Creates a group for human operators and attaches the MFA enforcement policy.
     */
    // Operators group and MFA enforcement removed.
    
    /**
     * Creates a strict password policy for IAM users.
     * Enforces 90-day rotation, complexity requirements, and prevents password reuse.
     */
    
    /**
     * Creates an IAM role for S3 read-only access with least privilege principle.
     */
    private Role createS3ReadOnlyRole() {
        return Role.Builder.create(this, EnvironmentConfig.getResourceName("iam", "s3-readonly-role"))
                .assumedBy(new ServicePrincipal("ec2.amazonaws.com"))
                .description("Role for EC2 instances requiring read-only S3 access")
                .inlinePolicies(Map.of(
                    "S3ReadOnlyPolicy", PolicyDocument.Builder.create()
                        .statements(List.of(
                            PolicyStatement.Builder.create()
                                .effect(Effect.ALLOW)
                                .actions(List.of(
                                    "s3:GetObject",
                                    "s3:GetObjectVersion",
                                    "s3:ListBucket"
                                ))
                                .resources(List.of(
                                    String.format("arn:aws:s3:::%s-*", EnvironmentConfig.ENVIRONMENT),
                                    String.format("arn:aws:s3:::%s-*/*", EnvironmentConfig.ENVIRONMENT)
                                ))
                                .build()
                        ))
                        .build()
                ))
                .build();
    }
    
    /**
     * Creates an IAM role for CloudTrail service with minimal required permissions.
     */
    private Role createCloudTrailRole() {
        return Role.Builder.create(this, EnvironmentConfig.getResourceName("iam", "cloudtrail-role"))
                .assumedBy(new ServicePrincipal("cloudtrail.amazonaws.com"))
                .description("Role for CloudTrail logging service")
                .inlinePolicies(Map.of(
                    "CloudTrailLogsPolicy", PolicyDocument.Builder.create()
                        .statements(List.of(
                            PolicyStatement.Builder.create()
                                .effect(Effect.ALLOW)
                                .actions(List.of(
                                    "logs:CreateLogGroup",
                                    "logs:CreateLogStream",
                                    "logs:PutLogEvents",
                                    "logs:DescribeLogGroups",
                                    "logs:DescribeLogStreams"
                                ))
                                // Limit to CloudWatch Logs in this account; CloudTrail writes to log groups with the pattern /aws/cloudtrail/*
                                .resources(List.of(String.format("arn:aws:logs:*:%s:log-group:/aws/cloudtrail/*", software.amazon.awscdk.Stack.of(this).getAccount())))
                                .build()
                        ))
                        .build()
                ))
                .build();
    }

    /**
     * Create an account password policy using the L1 construct so the account
     * enforces 90-day rotation and complexity rules.
     */
    private void createPasswordPolicy() {
        // Use an AWS SDK custom resource to call PutAccountPasswordPolicy because
        // CloudFormation does not recognize an account-level AccountPasswordPolicy resource.
        AwsSdkCall sdkCall = AwsSdkCall.builder()
            .service("IAM")
            // Use the service operation name expected by the CDK custom resource runtime
            // (some runtimes expect the PascalCase operation name).
            .action("PutAccountPasswordPolicy")
            .parameters(java.util.Map.of(
                "MinimumPasswordLength", EnvironmentConfig.PASSWORD_MIN_LENGTH,
                "RequireSymbols", EnvironmentConfig.REQUIRE_SYMBOLS,
                "RequireNumbers", EnvironmentConfig.REQUIRE_NUMBERS,
                "RequireUppercaseCharacters", EnvironmentConfig.REQUIRE_UPPERCASE,
                "RequireLowercaseCharacters", EnvironmentConfig.REQUIRE_LOWERCASE,
                "MaxPasswordAge", EnvironmentConfig.PASSWORD_MAX_AGE_DAYS,
                "PasswordReusePrevention", EnvironmentConfig.PASSWORD_REUSE_PREVENTION
            ))
            .physicalResourceId(PhysicalResourceId.of(EnvironmentConfig.getResourceName("iam", "password-policy")))
            .build();

        AwsCustomResource.Builder.create(this, EnvironmentConfig.getResourceName("iam", "password-policy"))
            .onCreate(sdkCall)
            .onUpdate(sdkCall)
            .policy(AwsCustomResourcePolicy.fromSdkCalls(SdkCallsPolicyOptions.builder()
                .resources(java.util.List.of("*"))
                .build()))
            .build();
    }

    /**
     * Creates a managed policy that can be attached to user groups to enforce MFA
     * and limit actions for human users. This is intentionally conservative.
     */
    private ManagedPolicy createMfaEnforcementPolicy() {
    // MFA enforcement removed per request.
    return null;
    }
    
    // Getters
    public Role getS3ReadOnlyRole() { return s3ReadOnlyRole; }
    public Role getCloudTrailRole() { return cloudTrailRole; }
    // Operators group removed.
    // Password policy getter intentionally omitted; refer to account-level governance
}